src/store/appReducer.js

```diff
@@ -0,0 +1,84 @@
+import { getUKFormattedDate } from '../utils/helpers';
+
+const appReducer = (state, action) => {
+  switch (action.type) {
+    case 'RESTORE_EXPENSES':
+      return { ...state, expenses: action.payload };
+
+    case 'ADD_EXPENSE_SUCCESS':
+      const expense = action.payload;
+      const expenseMonth = getUKFormattedDate(expense.date, { year: 'numeric', month: 'long' });
+
+      return {
+        ...state,
+        lastAddedExpense: expense,
+        isConfirmMonthModalOpen: expenseMonth !== state.selectedMonth,
+        didErrorOccur: false,
+        isAlertOpen: true,
+        userAction: 'add_expense',
+      };
+
+    case 'ADD_EXPENSE_FAIL':
```

Comment:
can we generalize the fail case, so we do not have to have two different `ADD_EXPENSE` action types?

currently, the answer might be yes, because we need to set `userAction`. is there a way to eliminate `userAction`?

Reply:
If we get rid of `userAction`, we will need to reintroduce individual `isOpen` state variables for each alert type. Right now we are using a single `isAlertOpen` variable in combination with `userAction` and `didErrorOccur` to trigger success/fail notifications for adding, deleting and editing.

Reply:
alrighty, i understand. i think there would be some value in re-opening this issue in another merge request (along with finding a solution for modals). i think trying out a one or two "other ways" may teach us some lessons and lead to enlightenment. (this is just a hunch; i have no concrete solution in my head)

Reply:
Maybe something like [this](https://jasonwatmore.com/post/2020/04/11/react-hooks-bootstrap-alert-notifications)?

> ...trigger alert notifications from any component by simply importing the alert service and calling one of it's methods for displaying different types of bootstrap alerts